src/wasm/activity-service/activity/lap.go

```diff
@@ -67,30 +66,51 @@ func NewLapFromRecords(records []*Record, sport string) *Lap {
 		heartRateAccumu.Collect(rec.HeartRate)
 		powerAccumu.Collect(rec.Power)
 		temperatureAccumu.Collect(rec.Temperature)
+	}
 
-		if i == 0 {
+	// Calculate Total Elapsed and Total Moving Time
+	for i := 0; i < len(records); i++ {
+		rec := records[i]
+		if rec.Timestamp.IsZero() {
 			continue
 		}
 
-		prev := records[i-1]
-
-		// Calculate Total Elapsed and Total Moving Time
-		if rec.Distance != nil && prev.Distance != nil {
-			timeDiff := rec.Timestamp.Sub(prev.Timestamp).Seconds()
-			lap.TotalElapsedTime += timeDiff
+		// Find next non-zero timestamp
+		for j := i + 1; j < len(records); j++ {
+			next := records[j]
+			if !next.Timestamp.IsZero() {
+				delta := next.Timestamp.Sub(rec.Timestamp).Seconds()
+				lap.TotalElapsedTime += delta
 
-			if IsConsideredMoving(sport, rec.Speed) {
-				lap.TotalMovingTime += timeDiff
+				if IsConsideredMoving(sport, rec.Speed) {
+					lap.TotalMovingTime += delta
+				}
+				i = j - 1 // move cursor
+				break
 			}
 		}
+	}
+
+	// Calculate Total Ascent and Total Descent
+	var totalAscent, totalDescent float64
+	for i := 0; i < len(records)-1; i++ {
+		rec := records[i]
+		if rec.SmoothedAltitude == nil {
+			continue
+		}
 
-		// Calculate Total Ascent and Total Descent
-		if rec.Altitude != nil && prev.Altitude != nil {
-			delta := *rec.Altitude - *prev.Altitude
-			if delta > 0 {
-				totalAscent += delta
-			} else {
-				totalDescent += math.Abs(delta)
+		// Find next non-nil altitude
+		for j := i + 1; j < len(records); j++ {
+			next := records[j]
+			if next.SmoothedAltitude != nil {
+				delta := *next.SmoothedAltitude - *rec.SmoothedAltitude
```

Comment:
update, should have use SmoothedAltitude for better result, this field is what we return to FE as `altitude` as well, so we get same data. 

Note: the `Altitude` field in WASM is intended for encoding so we encode raw value.

Reply:
then `SmoothedAltitude` is already `altitude` when returned to FE/JS. so FE/JS can directly compare without smoothed again

Reply:
Yes, correct